flik/linesearch/linesearch.py

```diff
@@ -0,0 +1,76 @@
+
+from flik.linesearch.tools import check_input
+from flik.linesearch.interpolations import cubic_interpolation
+from flik.linesearch.conditions import wolfe, armijo
+
+
+def line_search_general(var, func, grad, direction, alpha, conditions,
+                        const1=1e-4, const2=1e-2):
+    r"""Find a next-step alpha.
+
+    Parameters
+    ----------
+    var: {np.array(N,), float, int}
+        Variables' current values.
+    func: callable
+        Function :math:`f(x)`.
+    grad: callable
+        Gradient of the function :math:`\nabla f(x)`.
+    direction: {np.ndarray(N,), float, int}
+        Direction vector for next step.
+    alpha: float
+        The current step length estimate
+    conditions: str or list of str
+        Conditions to be satisfied by the new step length.
+        Valid options are: 'soft-wolfe', 'strong-wolfe' and 'armijo'
+    const1: float
+        Constant c_1 for Wolfe/Armijo conditions
+    const2: float
+        Constant c_2 for Wolfe conditions
+
+    Returns
+    -------
+    alpha_new: float
+        The optimal :math:`\alpha` of :math:`f(x + \alpha p)` that satisfies
+        the conditions requested.
+
+    """
+    # Check input quality
+    check_input(var=var, grad=grad, direction=direction, alpha=alpha)
+    if isinstance(conditions, str):
+        conditions = [conditions]
+    if not isinstance(conditions, list):
+        raise TypeError("The conditions should be given as a str or a list of str.")
+    if not isinstance(const1, float):
+        raise TypeError("Constant c_1 should be given as a float.")
+    if not isinstance(const2, float):
+        raise TypeError("Constant c_2 should be given as a float.")
+    if not 0 < const1 < const2 < 1.0:
+        raise ValueError("Constants c_1 and c_2  must meet the condition:"
+                         " 0 < c_1 < c_2 < 1.0")
+    # Check duplicates if there are any
+    if len(conditions) > 1:
+        con = [c for n, c in enumerate(conditions) if c not in conditions[:n]]
+        if con != conditions:
+            raise ValueError("One or more conditions are duplicated.")
+    # Check valid conditions
+    for i, condition in enumerate(conditions):
+        if condition not in ['soft-wolfe', 'strong-wolfe', 'armijo']:
+            raise ValueError("Condition %d+1 is not implemented." % i)
+
+    # Interpolate until alpha_new satisfies the conditions
+    alpha_new = 1.0
+    while True:
+        alpha_new = cubic_interpolation(var, func, grad, direction, alpha, alpha_new)
+        for condition in conditions:
+            if 'wolfe' in condition:
+                if 'soft' in condition:
+                    strong = False
+                else:
+                    strong = True
+                if wolfe(grad, var, direction, alpha_new, strong_wolfe=strong):
+                    return alpha_new
+            elif condition == 'armijo':
+                if armijo(func, grad, var, direction, alpha_new):
+                    return alpha_new
```

Comment:
The algorithm doesn't quite do what it says it'll do. It doesn't make sure that all of the conditions are satisfied. It just returns the first alpha that satisfies one of the conditions.